### PR TITLE
54 Nested properties validation

### LINKED ISSUE #54: Lack of resource validation for nested properties
We have received the report that are not validating nested properties in resources e.g.

```  
SomeDistribution:
    Type: AWS::CloudFront::Distribution
    Properties:
      DistributionConfig:
        Origins:
        ...
```

Validator will confirm that `DistributionConfig` is present - it’s a required property, but it will not dive into the properties of `DistributionConfig`.

### PR DESCRIPTION
**Pull request by mwpolcik**
This closes #54.

### CODE REVIEW

File: logger/logger.go [hunk 1]
```diff
@@ -117,10 +122,37 @@ func (logger *Logger) log(verbosity Verbosity, message string) {
...
+					fmt.Println("        ", err)
+				}
+			} else {
+				fmt.Println(resourceValidation.ResourceName, " has no validation errors")
+			}
+		}
+	}
+}
+
+// HasValidationErrors : Check if logger contains error messages
```
Comment (afronski) [inline logger/logger.go]: I believe we do not need function name and a colon here for documentation purposes.

File: .gitignore [hunk 2]
```diff
@@ -20,3 +20,6 @@ perun
 
 # Local perun configuration:
 .perun
+preprocessed.yml
```
Comment (afronski) [inline .gitignore]: Out of curiosity: what is it?

File: .gitignore [hunk 3]
```diff
@@ -20,3 +20,6 @@ perun
...
+
+.vscode
```
Comment (afronski) [inline .gitignore]: Let's mark it with a comment (what are those files and from which tool).

**Comment by mwpolcik** [comment]
Few comments from me:
1. It is a little bit of spaghetti code, so please be careful during review :)
2. Type Map property has limited validation, because most of the item types of maps are PrimitiveTypes(String, Bool). There is only one exception to this which was not covered.
3. I am not sure if this is covered in the commend in fixMultiLineMap function in fixMultiLineMap.go, but this:
```
		split := strings.Split(*line, name)
		if strings.Contains(pLines[idx+1], "-") && split[1] == "" {
```
In most cases  throw an error: **Index out of range**, e.g.
```
      SecurityGroups:
        - !FindInMap [ EnvSettings, !Ref "ENV", SecurityGroupId ]
```

File: .gitignore (hunk 2 above)
Comment (mwpolcik) [reply .gitignore]: This is a temporary file which is being created when `FixFunctions` is called from `intrinsicsolver` package.

**Comment by afronski** [comment]
@jlampar have a look on what @mwpolcik posted here, I think that bit is related to the `instrinsicsolver.go`. Also, I'd love to avoid creating a temporary file if possible (`preprocessed.yml`).

**Comment by jlampar** [comment]
I investigated this and enclosing everything in `if idx+1 < len(pLines)` condition does the trick. I'll be updating `instrinsicsolver.go` soon because I am using it for conversion improvements (changing short names to long). This change will be included.
Creating `preprocessed.yml` was proposed for two reasons:
1. To have preprocessed file to investigate any failures in further parsing (if preprocessing is going well, maybe the original template is noted in a strange manner?).
2. To have any solution at the moment. I haven't figured it yet how writing lines to []byte variable would be performed.

**Comment by afronski** [comment]
@jlampar, In that case, please prefix `preprocessed.yml` with `.` in order to hide that file on _UNIX_ / _Linux_ based systems.

**Pull request merged**